### PR TITLE
Use 32bit for indexing in `multi_tensor_apply` as possible

### PR DESCRIPTION
**Pull request by crcrpar**
I do think there's a better way to propagate what type to use for indexing, but I found an anonymous argument convenient to some extent as I was writing. Also, I can imagine the availability check of 32bit is a bit too optimistic.

Rel:
- https://github.com/pytorch/pytorch/pull/101760

---

Used optimizer with huggingface/transformers gpt2 setting as it doesn't need int64_t for indexing.
The second row is to update `state_step` tensors of 1 element.

**This PR**

```
 Time (%)  Total Time (ns)  Instances  Avg (ns)  Med (ns)  Min (ns)  Max (ns)  StdDev (ns)                                                  Name                                                
 --------  ---------------  ---------  --------  --------  --------  --------  -----------  ----------------------------------------------------------------------------------------------------
     99.2        341260837       1000  341260.8  529094.0      7071    542342     240433.6  void at::native::<unnamed>::multi_tensor_apply_kernel<at::native::<unnamed>::FusedOptimizerTensorLi…
      0.2           665096        200    3325.5    3328.0      3072      3616        107.8  void at::native::<unnamed>::multi_tensor_apply_kernel<at::native::<unnamed>::TensorListMetadata<(in…
```

**Main (`2.1.0a0+git07104ca`)**
```
 Time (%)  Total Time (ns)  Instances  Avg (ns)  Med (ns)  Min (ns)  Max (ns)  StdDev (ns)                                                  Name
 --------  ---------------  ---------  --------  --------  --------  --------  -----------  ----------------------------------------------------------------------------------------------------
     99.2        341346490       1000  341346.5  530021.0      7104    541893     240468.5  void at::native::<unnamed>::multi_tensor_apply_kernel<at::native::<unnamed>::FusedOptimizerTensorLi…
      0.2           664935        200    3324.7    3343.5      3072      3585        110.9  void at::native::<unnamed>::multi_tensor_apply_kernel<at::native::<unnamed>::TensorListMetadata<(in…
```

script

```python
import torch
from accelerate import Accelerator, DistributedType
from accelerate.logging import get_logger
from accelerate.utils import set_seed
from datasets import load_dataset
from huggingface_hub import Repository, create_repo
from torch.utils.data import DataLoader
from tqdm.auto import tqdm

import transformers
from transformers import (
    CONFIG_MAPPING,
    AutoModelForCausalLM,
)


config = CONFIG_MAPPING["gpt2"]()
model = AutoModelForCausalLM.from_config(config).cuda()

no_decay = ["bias", "layer_norm.weight"]
optimizer_grouped_parameters = [
	{
		"params": [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)],
		"weight_decay": 0.1,
	},
	{
		"params": [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)],
		"weight_decay": 0.0,
	},
]
optimizer = torch.optim.AdamW(optimizer_grouped_parameters, fused=True)


with torch.no_grad():
	for p in model.parameters():
		if p.requires_grad:
			p.grad = torch.rand_like(p)


for _ in range(100):
	optimizer.step()

print("DONE")
```

command
```console
nsys nvprof python script.py
```

cc @mcarilli

### CODE REVIEW

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh [hunk 1]
```diff
@@ -18,13 +19,13 @@ inline void increment_version(TensorList tensors) {
...
+template <int depth, typename T, typename index_t>
 __device__ bool init_args(
     T** args,
     TensorListMetadata<depth>& tl,
-    const int64_t chunk_idx,
-    const int64_t chunk_size,
-    const int64_t tensor_loc) {
+    const index_t chunk_idx,
+    const int chunk_size,
+    const int tensor_loc) {
```
Comment (crcrpar) [inline aten/src/ATen/native/cuda/ForeachFunctors.cuh:28]: `chunk_size` is always the costexpr var of `kChunkSize` defined in `MultiTensorApply.cuh`. I find it worth thinking of removing this argument but it'd break any cuda extensions of multi_tensor_apply`.

`tensor_loc` IIUC is basically `unsigned char`.

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh [hunk 2]
```diff
@@ -95,15 +96,23 @@ struct FusedOptimizerTensorListMetadata {
   int start_tensor_this_launch;
 };
 
-template <typename T, typename U, typename... ArgTypes>
+template <typename T, typename U, typename index_t, typename... ArgTypes>
 C10_LAUNCH_BOUNDS_1(kBlockSize)
 __global__ void multi_tensor_apply_kernel(
     T tensorListMeta,
+    const index_t arg_of_index_t,
```
Comment (crcrpar) [inline aten/src/ATen/native/cuda/MultiTensorApply.cuh:103]: The value of this argument doesn't mean a lot, the type does. This is why I defined two constexpr variables below.

**Comment by janeyx99** [comment]
Is the intended impact of this PR perf only? If so, I agree having benchmarking results would be great

**Comment by crcrpar** [comment]
added numbers to the description

**Comment by albanD** [comment]
Hey @crcrpar thanks for the benchmarks!
I'm curious about these numbers. Is there any caching happening in these kernels? I'm surprised to see the min ~1us and the max being ~500us? Or are there multiple indepent calls here that get globed while they shouldn't? 

Also to get a sense on the perf improvement, should we compare the two mean runtimes? Did we expect to see a bigger improvement from this or is this the expected behavior?

**Comment by crcrpar** [comment]
closing for now as I haven't been able to see a convincing performance gain on A100 & some huggingface/transformers examples

**Pull request closed without merging**